packages/datadog-plugin-rhea/src/index.js

```diff
@@ -1,73 +1,92 @@
-        'amqp.link.role': 'sender',
-        'out.host': this.connection.options.host,
-        'out.port': this.connection.options.port
-      } }, (span, done) => {
-        msg.delivery_annotations = msg.delivery_annotations || {}
-        tracer.inject(span, 'text_map', msg.delivery_annotations)
-        const delivery = send.apply(this, arguments)
-        if (this.options.snd_settle_mode !== 1) {
-          delivery._dd = { done, span }
-        } else {
-          done()
+      let host
+      let port
+      if (this.connection && this.connection.options) {
+        if (this.connection.options.host !== undefined) { host = this.connection.options.host }
+        if (this.connection.options.port !== undefined) { port = this.connection.options.port }
+      }
+      if (!CircularBuffer && this.session && this.session.outgoing && this.session.outgoing.deliveries) {
+        CircularBuffer = this.session.outgoing.deliveries.constructor
+        patchCircularBuffer(CircularBuffer)
+      }
+      return tracer.trace('amqp.send', {
+        tags: {
+          'component': 'rhea',
+          'resource.name': name,
+          'service.name': config.service || `${tracer._service}-amqp-producer`,
+          'span.kind': 'producer',
+          'amqp.link.target.address': name,
+          'amqp.link.role': 'sender',
+          'out.host': host,
+          'out.port': port
+        }
+      }, (span, done) => {
+        if (msg) {
+          msg.delivery_annotations = msg.delivery_annotations || {}
+          tracer.inject(span, 'text_map', msg.delivery_annotations)
         }
+        const delivery = send.apply(this, arguments)
+        delivery._dd = { done, span }
         return delivery
       })
     }
   }
 }
 
 function createWrapConnectionDispatch (tracer, config) {
+  function addTags (entry, error) {
+    if (entry && entry._dd) {
+      const { span, done } = entry._dd
+      if (error) {
+        span.addTags({ error })
+      }
+      done()
+      delete entry._dd
+    }
+  }
   return function wrapDispatch (dispatch) {
     return function dispatchWithTrace (eventName, obj) {
-      if (eventName === 'disconnected') {
+      if (eventName === 'disconnected' && this.local_channel_map) {
+        const error = obj.error || this.saved_error
         for (const key in this.local_channel_map) {
           const session = this.local_channel_map[key]
-          const addTags = entry => {
-            if (entry && entry._dd) {
-              const { span, done } = entry._dd
-              const error = obj.error || this.saved_error
-              if (error) {
-                span.addTags({
-                  'error.type': error.name,
-                  'error.msg': error.message,
-                  'error.stack': error.stack
-                })
-              }
-              done()
-            }
+          if (session && session.incoming && session.incoming.deliveries && session.incoming.deliveries.entries) {
```

Comment:
This block and the below block are almost identical and could be extracted to a function.